src/hardware/gus.cpp

```diff
@@ -1,924 +1,1418 @@
+// normalize all the volume index variables (including this) by multiplying by
+// VOLUME_INC_SCALAR (or 512). Note that "index" qualifies all these variables
+// because they are merely indexes into the vol_scalars[] array. The actual
+// volume scalar value (a floating point fraction between 0.0 and 1.0) is never
+// actually operated on, and is simply looked up from the final index position
+// at the time of sample population.
+void Voice::WriteVolRate(uint16_t val)
+{
+	vol_ctrl.rate = val;
+	constexpr uint8_t bank_lengths = 63;
+	const int pos_in_bank = val & bank_lengths;
+	const int decimator = 1 << (3 * (val >> 6));
+	vol_ctrl.inc = ceil_sdivide(pos_in_bank * VOLUME_INC_SCALAR, decimator);
+
+	// Sanity check the bounds of the incrementer
+	assert(vol_ctrl.inc >= 0 && vol_ctrl.inc <= bank_lengths * VOLUME_INC_SCALAR);
+}
+
+void Voice::WriteWaveRate(uint16_t val)
+{
+	wave_ctrl.rate = val;
+	wave_ctrl.inc = ceil_udivide(val, 2u);
+}
+
+Gus::Gus(uint16_t port, uint8_t dma, uint8_t irq, const std::string &ultradir)
+        : port_base(port - 0x200),
+          dma2(dma),
+          irq1(irq),
+          irq2(irq)
+{
+	// Create the internal voice channels
+	for (uint8_t i = 0; i < MAX_VOICES; ++i) {
+		voices.at(i) = std::make_unique<Voice>(i, voice_irq);
 	}
-	INLINE void WaveUpdate(void) {
-		if (WaveCtrl & ( WCTRL_STOP | WCTRL_STOPPED)) return;
-		Bit32s WaveLeft;
-		if (WaveCtrl & WCTRL_DECREASING) {
-			WaveAddr -= WaveAdd;
-			WaveLeft = WaveStart-WaveAddr;
-		} else {
-			WaveAddr += WaveAdd;
-			WaveLeft = WaveAddr-WaveEnd;
-		}
-		//Not yet reaching a boundary
-		if (WaveLeft<0) 
-			return;
-		/* Generate an IRQ if needed */
-		if (WaveCtrl & 0x20) {
-			myGUS.WaveIRQ|=irqmask;
-		}
-		/* Check for not being in PCM operation */
-		if (RampCtrl & 0x04) return;
-		/* Check for looping */
-		if (WaveCtrl & WCTRL_LOOP) {
-			/* Bi-directional looping */
-			if (WaveCtrl & WCTRL_BIDIRECTIONAL) WaveCtrl^= WCTRL_DECREASING;
-			WaveAddr = (WaveCtrl & WCTRL_DECREASING) ? (WaveEnd-WaveLeft) : (WaveStart+WaveLeft);
-		} else {
-			WaveCtrl|=1;	//Stop the channel
-			WaveAddr = (WaveCtrl & WCTRL_DECREASING) ? WaveStart : WaveEnd;
-		}
+
+	RegisterIoHandlers();
+
+	// Register the Audio and DMA callbacks
+	audio_channel = mixer_channel.Install(
+		std::bind(&Gus::AudioCallback, this, std::placeholders::_1), 1, "GUS");
+
+	UpdateDmaAddress(dma);
+
+	// Populate the volume, pan, and auto-exec arrays
+	PopulateVolScalars();
+	PopulatePanScalars();
+	PopulateAutoExec(port, ultradir);
+}
+
+void Gus::ActivateVoices(uint8_t requested_voices)
+{
+	requested_voices = clamp(requested_voices, MIN_VOICES, MAX_VOICES);
+	if (requested_voices != active_voices) {
+		active_voices = requested_voices;
+		assert(active_voices <= voices.size());
+		active_voice_mask = 0xffffffffU >> (MAX_VOICES - active_voices);
+		playback_rate = static_cast<uint32_t>(
+		        0.5 + 1000000.0 / (1.619695497 * active_voices));
```

Comment:
Suggestion: This seems like a naive rounding up; maybe you can use `ceil` function in here instead?

Reply:
Ceil is correct in some cases, but off-by-one in others:

https://onlinegdb.com/rklTz8l5MP

![2020-08-18_17-44](https://user-images.githubusercontent.com/1557255/90579331-89a87600-e17a-11ea-925d-e495d952aab8.png)

Reply:
Round pulls the answer in the right direction (avoiding the 0.5 to act as the truncating tie-breaker):

https://onlinegdb.com/SJ_5dlqMD

![2020-08-18_17-56](https://user-images.githubusercontent.com/1557255/90579859-1e5fa380-e17c-11ea-8307-cf330633ffb5.png)